### PR TITLE
Convert command sending to use archetype chunks

### PR DESCRIPTION
**Pull request by zeroZshadow**
#### Description
This is both an optimisation and redesign.

With this redesign, the command sending will do less iterations, and uses faster component iterators to improve overall performance of the SpatialOSSendSystem.
It is also a good example of how Archetype Chunks can be used in our core, and how they work with code generation.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 namespace Improbable.Gdk.Core
 {
     [DisableAutoCreation]
+    [AlwaysUpdateSystem]
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs]: we are not injecting anything so it should always run by default

File: workers/unity/Packages/com.improbable.gdk.core/CodegenAdapters/ComponentReplicationHandler.cs [hunk 2]
```diff
@@ -8,10 +8,9 @@ public abstract class ComponentReplicationHandler
     {
         public abstract uint ComponentId { get; }
         public abstract ComponentType[] ReplicationComponentTypes { get; }
-        public abstract ComponentType[] CommandTypes { get; }
 
         public abstract void ExecuteReplication(ComponentGroup replicationGroup, Connection connection);
-        public abstract void SendCommands(List<ComponentGroup> commandComponentGroups, Connection connection);
+        public abstract void SendCommands(SpatialOSSendSystem sendSystem, Connection connection);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/CodegenAdapters/ComponentReplicationHandler.cs]: This should just be `ComponentSystem` imo - no need to over constrain it

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs (hunk 1 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs]: Things are injected - `GetComponentGroup(...)` registers the same underlying things as marking a struct as `[Inject]`

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs [hunk 3]
```diff
@@ -82,10 +81,10 @@ private struct ComponentReplicator
             public ComponentGroup ReplicationComponentGroup;
             public List<ComponentGroup> CommandReplicationGroups;
 
-            public void Execute(Connection connection)
+            public void Execute(SpatialOSSendSystem sendSystem, Connection connection)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs]: Drop the first argument and pass in `this` directly into `SendCommands`

Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSSendSystem.cs]: no

File: test-project/Assets/Generated/Improbable.Gdk.Generated.asmdef [hunk 4]
```diff
@@ -1,9 +1,10 @@
 ﻿{
 	"name": "Improbable.Gdk.Generated",
 	"references": [
-		"Unity.Entities", 
-		"Unity.Mathematics", 
-		"Unity.Entities.Hybrid", 
+	    "Unity.Collections",
+		"Unity.Entities",
```
Comment (jessicafalk) [inline test-project/Assets/Generated/Improbable.Gdk.Generated.asmdef]: nit: whitespaces

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt [hunk 5]
```diff
@@ -580,90 +590,76 @@ namespace <#= qualifiedNamespace #>
                 }
             }
 
-            public override void SendCommands(List<ComponentGroup> commandComponentGroups, global::Improbable.Worker.Core.Connection connection)
+            public override void SendCommands(SpatialOSSendSystem sendSystem, global::Improbable.Worker.Core.Connection connection)
             {
+                var entityType = sendSystem.GetArchetypeChunkEntityType();
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: move that into a check as well. Right now we assign it in many generated files without actually using it

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt [hunk 6]
```diff
@@ -513,16 +514,25 @@ namespace <#= qualifiedNamespace #>
...
 
-            public override ComponentType[] CommandTypes => new ComponentType[] {
 <# foreach (var commandDetails in commandDetailsList) { #>
-                ComponentType.ReadOnly<<#= qualifiedNamespace #>.<#= componentDetails.ComponentName #>.CommandSenders.<#= commandDetails.CommandName #>>(),
-                ComponentType.ReadOnly<<#= qualifiedNamespace #>.<#= componentDetails.ComponentName #>.CommandResponders.<#= commandDetails.CommandName #>>(),
+            private CommandStorages.<#= commandDetails.CommandName#> <#= commandDetails.CamelCaseCommandName #>Storage;
 <# } #>
-            };
 
+            private EntityArchetypeQuery[] CommandQueries =
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: Does this need to be an array if we are unwrapping it in the SendCommands function? Is it used somewhere else?

Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: if we have multiple commands in a component, you will need a query for every type of command

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt (hunk 5 above)
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: Already did this in a follow up PR that also adds Profiling markers everywhere. Sorry that this is split up.

**Pull request merged**